### ISSUE TITLE
Intrinsic rewards

### ISSUE THREAD

**Issue by PeterZhizhin** [tensorflow/agents#21]
Thank you for such an amazing work, it is interesting to see what TF-Agents has to offer!

It would be great if we had an ability to encourage exploration with intrinsic motivators.
I would like to see Exploration by Random Network Distillation (https://arxiv.org/abs/1810.12894) as an example of such motivators.

It would be interesting to see your ideas how one could implement it. It should be a somewhat general mechanism, not related to any specific policy implementation.

Original implementation of RND uses PPO with the following modifications of the policy:
1. Intrinsic reward is added. It is computed as a distance between predicted features and target features. Target features are computed as a randomly-initialized fixed neural network on a next observation after an action. Predicted features is a neural network that outputs a tensor of the same length as a target network. After each sampling episode a fraction of experience is used to update the predictor's network weights.
2. There are two critics, not one. The first one predicts extrinsic value function, the second predicts intrinsic value function. Advantage is computed separately using both models.
3. Advantage in PPO should be a weighted combination of extrinsic and intrinsic critics predictions.

For Deep Q-learning, it should also be similar.
1. We need another network that predicts Q function for intrinsic reward that is used only during training.
2. Sampling strategy should be modified to use the second Q-function. It should sample results based on a total Q function that is computed as a weighted sum of two other ones.

We can continue the list for other algorithms. I would like to see your ideas how I could implement RND using TF-Agents in a way that it can be used in multiple policies. I also want to know whether you would accept such contribution if I made it.

**Comment by samfishman** [comment]
Hi Peter,

Thanks for your comment and distillation of the work required to implement RND. I think that the cleanest way to implement this would be something along the following lines.

Computing the intrinsic reward and training the RND networks can be done mostly agent-independently:
1. Implement an environment wrapper (see [here](https://github.com/tensorflow/agents/blob/master/tf_agents/environments/wrappers.py#L40-L67)) to compute the intrinsic reward using the RND target and predictor networks. You can expose this as an `intrinsic_reward` property on a subclass of the `TimeStep` object.

2. Expose a `train` method on the environment wrapper to take in a batch of trajectories and update the predictor method parameters based on the RND loss.

3. Modify the train_eval script to train the RNDWrapper as well as the agent.

I think modifying the agent to use the intrinsic reward will unfortunately have to be a little agent-specific since the method involves 2 different Q or Value networks. For PPO, it might look like this:
1. Subclass the PPOAgent.
2. In the subclassed constructor, `.copy()` the value network to make a second intrinsic reward value network.
3. Extract the logic of `compute_return_and_advantage` to a helper function that takes in the `reward` tensor and a value network as a parameter.
4. Override `compute_return_and_advantage` in your subclassed agent to call into the helper function once for the regular `reward` and once for `intrinsic_reward`, combining them using a weighted sum and returning that.

For DDPG, you would do something similar for the `critic_loss` function (calling it once with external reward and once with intrinsic reward).

We accept contributions of new algorithms if it is significant in the community and there are 2 Googlers willing to maintain the algorithm (which I can't make any promises for). If that doesn't happen, then we can link an implementation in your own repo from our README, which should still give it some good visibility and be a super useful contribution to the community. So I'd definitely encourage you to give a shot at implementing it!

**Comment by kbanoop** [comment]
Closing this issue, please reopen if you have further questions.
